### PR TITLE
 ONNX Export IsNan op

### PR DESCRIPTION
**Pull request by lara-hdr**

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_caffe2.py [hunk 1]
```diff
@@ -1057,6 +1057,14 @@ def forward(self, x):
         self.run_model_test(RsubModel(), train=False, input=(x,),
                             batch_size=BATCH_SIZE, use_gpu=False)
 
+    def test_isnan(self):
+        class IsNaNModel(torch.nn.Module):
+            def forward(self, input):
+                return torch.isnan(input)
+
+        x = torch.tensor([1.0, float('nan'), 2.0])
+        self.run_model_test(IsNaNModel(), train=False, input=x, batch_size=BATCH_SIZE)
```
Comment (houseroad) [inline test/onnx/test_pytorch_onnx_caffe2.py]: Please add `use_gpu=False` here.

File: caffe2/operators/utility_ops.cc [hunk 2]
```diff
@@ -833,6 +833,26 @@ OPERATOR_SCHEMA(NanCheck)
         "Tensor to copy input into if no NaNs or inf."
         " Can be in-place");
 
+template <>
+bool IsNanOp<CPUContext>::RunOnDevice() {
+  auto& X = Input(0);
+  auto* Y = Output(0, X.sizes(), at::dtype<uint8_t>());
+  const float* X_data = X.template data<float>();
```
Comment (houseroad) [inline caffe2/operators/utility_ops.cc]: We should assert it's `float` type. Cannot direct convert without checking.

Another option: create a template for data type as well. How to do that, you can check other op's implementation. :-)

File: caffe2/operators/utility_ops.h [hunk 3]
```diff
@@ -42,6 +42,34 @@ struct GetNanCheckGradient : public GradientMakerBase {
...
+    const auto* X_data = X.template data<T>();
+    uint8_t* Y_data = Y->template mutable_data<uint8_t>();
+    for (size_t i = 0; i < X.numel(); i++) {
+      Y_data[i] = (uint8_t)(std::isnan(X_data[i]));
+    }
+    return true;
+  }
+
+ private:
+  TensorPrinter tensorPrinter_;
```
Comment (houseroad) [inline caffe2/operators/utility_ops.h]: remove these unused fields?

**Comment by houseroad** [comment]
@lara-hdr to pass the CI, you need to rebase to a newer commit on master

**Review by houseroad — COMMENTED** [review]
Looks good, but please rebase the PR, and trigger the CI again.

**Comment by houseroad** [comment]
Oh, this is a nice feature !

**Comment by lara-hdr** [comment]
@houseroad all tests are passing with the rebase

**Comment by lara-hdr** [comment]
@houseroad , can we merge this? :)

**Review by houseroad — APPROVED** [review]
Looks good to me. Thanks!

**Review by zrphercule — APPROVED** [review]
LGTM, great job!

**Pull request closed without merging**